### PR TITLE
rewrite fetchos function to use pipeline

### PR DESCRIPTION
**Pull request by rcooke-warwick**
I have been trying to make the jetson-tx2 device type work consistantly with the os test suite.

The final issue I encountered was the HUP suite https://github.com/balena-os/meta-balena/tree/master/tests/suites/hup failing every single time. The problem was that the jetson would never flash its internal storage, when we try to provision it with the latest balena OS prod release. 

- the os and cloud suites worked fine when used with the same OS image version
- after some examination, the only difference I could see, was that in the os and cloud suites, we provided this image (I downloaded it via the browser), but in the hup suite, we use the leviathan core `fetchOS()` function to get this image.
- This threw the `fetchOS()` function into a suspicious light. So what I did, was tried the hup suite, instead using the CLI to download the image - and it worked all 3 times I tried it! 
- I then went into `fetchOS()` and experimented with rewriting parts of it. I removed the retry as it was never being used anyway, as far as I could see. I also used the `pipeline` utility to ensure correct handling of the read and write streams - as I suspect that something strange was happening with the previous imlpementation
- I'm not exactly sure why this worked - but it did seem to fix my problem. As to why this issue never appeared for other device types, I can only theorise that its because the jetson image is much larger (2.4G compared to the ~950M or the pi's, and 1.6G for NUC and rockpi)

Change-type: patch
Signed-off-by: Ryan Cooke <ryan@balena.io>

### CODE REVIEW

**Comment by rcooke-warwick** [comment]
@balena-ci rebase

**Comment by rcooke-warwick** [comment]
@balena-ci rebase

**Comment by klutchell** [comment]
> @balena-ci rebase

This is not available now that we disabled VersionBot.

Instead select the Update Branch dropdown and choose Update with Rebase.

File: core/lib/components/balena/sdk.js [hunk 1]
```diff
@@ -353,43 +354,46 @@ module.exports = class BalenaSDK {
...
+				async function(error, stream) {
+					if (error) {
+						fs.unlink(path, () => {
+							// Ignore.
 						});
-					},
-				);
-			});
-		};
-		return retry(downloadLatestOS, { max_tries: 3, interval: 500 });
```
[the remark this answers is not in the document]
Comment (vipulgupta2048) [reply core/lib/components/balena/sdk.js:393]: I just remembered. When I was checking it last, I rewrote it in TS

File: core/lib/components/balena/sdk.js [hunk 2]
```diff
@@ -353,43 +354,46 @@ module.exports = class BalenaSDK {
...
-									`Downloading balenaOS image: ${toInteger(data.percentage) + '%'
-									}`,
-								);
-								progress = data.percentage;
-							}
-						});
-						stream.pipe(fs.createWriteStream(path));
-						stream.on('finish', () => {
-							console.log(`Download Successful: ${path}`);
-							resolve(path);
```
Comment (vipulgupta2048) [inline core/lib/components/balena/sdk.js:387]: This would be helpful to put back in to mark the completion of the download

**Review by vipulgupta2048 — COMMENTED** [review]
Thanks for adding this in @rcooke-warwick Weird why the pipe is not able to handle this and we need pipeline? But good to see Jetson is unblocked now.

**Comment by rcooke-warwick** [comment]
@vipulgupta2048 I'm going to close this, as I don't think that it was actually the problem - further investigation led me to this fix instead: https://github.com/balena-os/meta-balena/pull/3022

**Pull request closed without merging**